File: aptu/604_17/koltsov/2/src/main/java/ru/au/parallel/LockFreeSet.java
package ru.au.parallel;

/**
 * Lock-Free множество.
 * @param <T> Тип ключей
 */
public interface LockFreeSet<T extends Comparable<T>> {
    /**
     * Добавить ключ к множеству
     *
     * Алгоритм должен быть как минимум lock-free
     *
     * @param value значение ключа
     * @return false если value уже существует в множестве, true если элемент был добавлен
     */
    boolean add(T value);

    /**
     * Удалить ключ из множества
     *
     * Алгоритм должен быть как минимум lock-free
     *
     * @param value значение ключа
     * @return false если ключ не был найден, true если ключ успешно удален
     */
    boolean remove(T value);

    /**
     * Проверка наличия ключа в множестве
     *
     * Алгоритм должен быть как минимум wait-free
     *
     * @param value значение ключа
     * @return true если элемент содержится в множестве, иначе - false
     */
    boolean contains(T value);

    /**
     * Проверка множества на пустоту
     *
     * Алгоритм должен быть wait-free (достаточно lock-free, wait-free для сильно уверенных в себе)
     *
     * @return true если множество пусто, иначе - false
     */
    boolean isEmpty();
}

File: aptu/302_18/smirnov/lock_free_set/src/main/kotlin/ru/spbau/mit/concurrent/set/LockFreeSet.kt
package ru.spbau.mit.concurrent.set

/**
 * Lock-Free множество.
 * @param <T> Тип ключей
 */
interface LockFreeSet<in T : Comparable<T>> {
    /**
     * Добавить ключ к множеству
     *
     * Алгоритм должен быть как минимум lock-free
     *
     * @param value значение ключа
     * @return false если value уже существует в множестве, true если элемент был добавлен
     */
    fun add(value: T): Boolean

    /**
     * Удалить ключ из множества
     *
     * Алгоритм должен быть как минимум lock-free
     *
     * @param value значение ключа
     * @return false если ключ не был найден, true если ключ успешно удален
     */
    fun remove(value: T): Boolean

    /**
     * Проверка наличия ключа в множестве
     *
     * Алгоритм должен быть как минимум lock-free
     *
     * @param value значение ключа
     * @return true если элемент содержится в множестве, иначе - false
     */
    fun contains(value: T): Boolean

    /**
     * Проверка множества на пустоту
     *
     * Алгоритм должен быть wait-free (достаточно lock-free, wait-free для сильно уверенных в себе)
     *
     * @return true если множество пусто, иначе - false
     */
    fun isEmpty(): Boolean
}